### DISCUSSION TITLE
Right Click OpenAI Menu

### DISCUSSION THREAD

**Discussion by slarrauri** [directus/directus#20227] [Feature Requests]
### Summary

The desired functionality in Directus allows users to right-click on selected text within an input field, which would trigger a menu with two options: using the selected text as a prompt or improving the selected text. Upon selecting either option, it would make an API call to OpenAI, and the response from the API would be utilized to replace or append beneath the selected text in the input field.

### Motivation

The motivation behind this feature is to enhance user experience and efficiency within Directus, as well as to support various use cases. The key constraints and objectives being addressed are:

1. **Improved User Interaction:** Allowing users to right-click on selected text and access options for utilizing that text provides a more intuitive and convenient interface. This can enhance the user experience by reducing the number of manual steps required to interact with selected text.

2. **Seamless Integration with OpenAI:** By enabling a direct interaction with the OpenAI API, Directus can leverage AI capabilities for text processing and enhancement. This supports use cases where users need to refine or expand upon selected text, for example, in content editing or generation.

3. **Productivity and Efficiency:** The feature aims to streamline text-related tasks by minimizing context switches and manual data input. Users can quickly utilize the selected text as a prompt for AI-driven improvements, which can save time and effort.

4. **Versatility:** The feature can accommodate various use cases beyond just text improvement. It could also support use cases where users need to extract insights, translation, or other text-related operations by interacting with the API.

5. **User-Friendly Interface:** Providing a menu for selecting the desired action ensures that users have clear choices when interacting with text. This enhances the user interface's clarity and accessibility.

The expected outcome is an improved user experience within Directus, increased efficiency in handling text-related tasks, and the ability to seamlessly integrate AI capabilities from OpenAI. By focusing on these motivations and constraints, alternative solutions can be explored if the specific feature proposal is not accepted, while still addressing the core needs of the user base. This approach allows for flexibility and adaptability in addressing the identified issues and opportunities.

### Detailed Design

Feature Name: Context Menu Integration for Text Enhancement

Design Overview:
This feature aims to enable users to right-click on selected text within an input field in Directus and access a context menu with two options: "Use as Prompt" and "Improve Text." Upon selecting one of these options, an API call will be made to OpenAI, and the response will be used to replace or append text in the input field. Below are the key components and considerations for implementing this feature:

User Interface Enhancement:

Context Menu: A right-click event on selected text in an input field triggers a context menu.
Menu Options: The context menu will contain two options: "Use as Prompt" and "Improve Text."
Integration with OpenAI:

API Configuration: Users should be able to configure their OpenAI API credentials within Directus to enable the integration.
API Calls: When a user selects either "Use as Prompt" or "Improve Text," an API call to OpenAI is made, passing the selected text as input.
Response Handling: The response from the OpenAI API is processed and used to either replace the selected text or append it below the selected text in the input field.
User Configuration:

Settings: Users should have the ability to customize settings such as which OpenAI model to use, API endpoint, and any other relevant parameters.
User Feedback: Users should receive feedback on the status of the API call, such as success or error messages, in a user-friendly manner.
Security and Privacy:

Secure API Usage: Ensure that API calls to OpenAI are made securely, with proper authorization and token handling.
User Data Privacy: Implement data handling and storage in compliance with privacy regulations and best practices.
Error Handling:

Connection Errors: Handle situations where the integration cannot establish a connection with the OpenAI API.
API Response Errors: Gracefully manage errors or unexpected responses from the OpenAI service.
Examples of Use:

Basic Use:

A user selects a block of text within an input field.
Right-clicking on the selected text reveals the context menu with options "Use as Prompt" and "Improve Text."
The user clicks on "Improve Text."
Directus makes an API call to OpenAI, and the response text is appended below the original text in the input field, providing improved content.
Customization:

Users can access settings to specify the OpenAI model to be used, API endpoints, and any other relevant parameters.
This allows users to tailor the text enhancement process according to their specific needs.
Error Handling:

If there is a connection error or an issue with the API call, users receive clear and informative error messages to understand and address the problem.
Terminology:

Context Menu: A user interface component that appears in response to a right-click action and provides a set of options relevant to the selected text.
OpenAI API: The interface to the OpenAI service, which offers various natural language processing capabilities.
Prompt: The initial text or input provided to the OpenAI service for generating improved or expanded content.

### Must-Have (M):

User Interface Integration (UI): The context menu must appear when a user right-clicks on selected text in an input field.
Menu Options: The context menu must contain two options: "Use as Prompt" and "Improve Text."
OpenAI API Integration: The feature must allow users to configure their OpenAI API credentials.
API Calls: When a user selects either "Use as Prompt" or "Improve Text," an API call to OpenAI must be made, passing the selected text as input.
Response Handling: The response from the OpenAI API must be processed and used to replace or append text in the input field.

### Should-Have (S):
6. User Configuration: Users should have the ability to customize settings, including the choice of OpenAI model, API endpoint, and other relevant parameters.

User Feedback: Users should receive clear and user-friendly feedback on the status of the API call (e.g., success or error messages).
Secure API Usage: API calls to OpenAI should be made securely, with proper authorization and token handling.
Error Handling: The feature should gracefully manage connection errors, as well as errors or unexpected responses from the OpenAI service.

### Could-Have (C):
10. Additional Menu Options: Consider adding more menu options or customization for users based on feedback and evolving use cases.

Integration with Third-Party Services: Explore the potential for integrating with other text enhancement or language processing services in addition to OpenAI.

### Won't-Have (W):
12. Offline Mode: An offline mode for this feature, which doesn't rely on external API services, is not considered a requirement at this time.

### Drawbacks

I don't know Directus well enough at the moment, to write about this. 

### Alternatives

Alternatives Considered:

Manual Copy-Paste: Users can manually copy the selected text, open the OpenAI platform separately, input the text, and then copy the enhanced text back into the Directus input field. This method is time-consuming and requires context switching between applications.

External Plugin: Instead of a built-in feature, users could rely on third-party browser extensions or plugins that integrate with Directus and OpenAI to provide similar functionality. However, this may lack consistency and control over user experience.

Text Enhancement as a Service: Directus could provide a built-in feature to send selected text to an external service specializing in text enhancement. This service could be OpenAI or another text enhancement provider. The impact of this choice would be similar to the proposed feature but may involve different integration and cost considerations.

Impact of Not Doing This:

Reduced Efficiency: Without this feature, users would need to manually switch between applications, copy and paste text, and manage the enhancement process separately. This results in reduced efficiency and increased manual effort.

Increased Room for Errors: Manual copying and pasting text between applications increases the likelihood of errors or inconsistencies in the text enhancement process, potentially impacting the quality of the output.

Higher Learning Curve: The absence of a built-in text enhancement feature can potentially increase the learning curve for new users who have to navigate third-party tools or manual processes.

Limited Control: Users may not have control over the quality of the text enhancement process when relying on external tools or manual steps. A built-in feature would provide more consistent and customizable results.

Inefficient Workflow: For users working with text intensively, the absence of this feature could lead to a less efficient and streamlined workflow, impacting productivity.

### Adoption Strategy

The implementation of this proposal does not constitute a breaking change and I'M unsure about whether a migration is necessary, and I don't have information about its potential impact on other projects in the Directus ecosystem, I should proceed with a more cautious and gradual approach.

**Comment by br41nslug** [comment]
Imho this is something more appropriate as extension than as core feature. Extensions integrating Directus with the OpenAI API already exist out there.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
